// bot-ts/dist/commands/unknown.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const model_1 = require("../model");
/**
 * Handles when the user enters a command that does not exist
 */
// tslint:disable-next-line: no-default-export
class UnknownCommand extends model_1.Command {
    constructor(client) {
        super(client, {
            name: 'unknown',
            description: 'Acts as a response to unknown commands.',
            group: 'info',
            unknown: true,
            default: true,
            args: [
                {
                    key: 'commandName',
                    phrase: 'What command are you wanting to run?',
                    type: 'string'
                }
            ]
        });
    }
    /**
     * Executes the command
     * @param msg - The object containing the user's message
     * @param args - Arguments passed to the command
     */
    async run(msg, args) {
        await msg.reply(`Unknown command \`${args.commandName}\`.`);
    }
}
exports.UnknownCommand = UnknownCommand;
