// bot-ts/dist/model/command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * A command for your bot.
 *
 * Give it some information, and control how it executes (and who can execute it).
 *
 * @example
 * ```ts
 * export class EchoCommand extends Command {
 *   constructor(client: Client) {
 *     super(client, {
 *       name: 'echo',
 *       description: 'Make me repeat something!',
 *       args: [
 *         {
 *           key: 'phrase',
 *           phrase: 'What should I echo?',
 *           type: 'string'
 *         }
 *       ]
 *     })
 *   }
 *
 *   public async run(msg: CommandMessage, args: {phrase: string}) {
 *     await msg.reply(args.phrase)
 *   }
 * }
 * ```
 */
class Command {
    constructor(client, options) {
        this.options = options;
        this.client = client;
    }
    /**
     * Find a command by name or alias.
     *
     * @param client - The parent client object
     * @param nameOrAlias - The name or alias of the command, e.g. `echo`
     */
    static find(client, nameOrAlias) {
        return (client.commands.find(x => x.options.name === nameOrAlias) ||
            this.findByAlias(client, nameOrAlias));
    }
    static findByAlias(client, alias) {
        return client.commands.find(x => !!x.options.aliases && x.options.aliases.includes(alias));
    }
    /**
     * Checks if the user has permission to run this command.
     *
     * @param msg - The message object
     */
    async hasPermission(msg) {
        return true;
    }
}
exports.Command = Command;
